### PR TITLE
ajax calls which return response with charset=utf-16 not parsed correctly

### PR DESCRIPTION
**Pull request by alvarorahul**

### CODE REVIEW

**Comment by alvarorahul** [comment]
#1398

**Comment by domenic** [comment]
Needs tests and needs to not break the build

**Comment by domenic** [comment]
Should probably be general over any encoding, not just utf-16.

File: lib/jsdom/living/xmlhttprequest.js [hunk 1]
```diff
@@ -590,6 +594,30 @@ module.exports = function createXMLHttpRequest(window) {
...
+      });
+    }
+
+    return result;
+  }
+
+  function bufferToString(buffer, contentType) {
+    const parsed = parseContentType(contentType);
+    let encoding;
+    if (parsed.params["charset"] === "utf-16") {
```
Comment (domenic) [inline lib/jsdom/living/xmlhttprequest.js]: Needs to compare lowercased

File: lib/jsdom/living/xmlhttprequest.js [hunk 2]
```diff
@@ -590,6 +594,30 @@ module.exports = function createXMLHttpRequest(window) {
     DONE: 4
   });
 
+  function parseContentType(contentType) {
+    const result = { type: "", params: { } };
+
+    if (contentType) {
+      const parts = contentType.split(";");
+      result.type = parts[0];
+      parts.slice(1).forEach(part => {
```
Comment (domenic) [inline lib/jsdom/living/xmlhttprequest.js]: Just use a `for (let i = 1; ...)` loop; no need to create a new array just to use `.forEach` on it.

File: lib/jsdom/living/xmlhttprequest.js [hunk 3]
```diff
@@ -590,6 +594,30 @@ module.exports = function createXMLHttpRequest(window) {
...
+        const kvp = part.split("=");
+        result.params[kvp[0].trim()] = kvp[1].trim();
```
Comment (domenic) [inline lib/jsdom/living/xmlhttprequest.js]: Need to remove quotes from potentially-quoted values (e.g. `Content-Type: text/html; charset="utf-8"`)

Comment (domenic) [inline lib/jsdom/living/xmlhttprequest.js]: Needs to handle cases where people to invalid things like `Content-Type: text/html; charset; foo; baz=bar=qux`.

File: lib/jsdom/living/xmlhttprequest.js [hunk 4]
```diff
@@ -590,6 +594,30 @@ module.exports = function createXMLHttpRequest(window) {
     DONE: 4
   });
 
+  function parseContentType(contentType) {
```
Comment (domenic) [inline lib/jsdom/living/xmlhttprequest.js]: Actually, this would all be a lot easier if maybe we just used https://www.npmjs.com/package/contenttype :))

Comment (alvarorahul) [inline lib/jsdom/living/xmlhttprequest.js]: agreed... will switch to use the contenttype lib

Comment (domenic) [inline lib/jsdom/living/xmlhttprequest.js]: https://www.npmjs.com/package/con-type looks pretty good too; maybe it is better at handling invalid cases.

File: lib/jsdom/living/xmlhttprequest.js (hunk 1 above)
Comment (alvarorahul) [inline lib/jsdom/living/xmlhttprequest.js]: switched to use https://www.npmjs.com/package/contenttype

File: lib/jsdom/living/xmlhttprequest.js (hunk 2 above)
Comment (alvarorahul) [inline lib/jsdom/living/xmlhttprequest.js]: switching to https://www.npmjs.com/package/contenttype

File: lib/jsdom/living/xmlhttprequest.js (hunk 3 above)
Comment (alvarorahul) [inline lib/jsdom/living/xmlhttprequest.js]: switching to https://www.npmjs.com/package/contenttype

File: lib/jsdom/living/xmlhttprequest.js (hunk 4 above)
Comment (alvarorahul) [inline lib/jsdom/living/xmlhttprequest.js]: how about this one https://www.npmjs.com/package/content-type?

**Comment by Sebmaster** [comment]
> Should probably be general over any encoding, not just utf-16.

Is there a more recent specification than the [IANA Charset spec](http://www.iana.org/assignments/character-sets/character-sets.xhtml) on which charsets exist / how to parse them?

I don't think the current switch case with hex/binary/etc. is a good thing, it probably produces a lot of invalid results.

**Comment by Sebmaster** [comment]
Found it, we should strive to support these encodings and aliases: https://encoding.spec.whatwg.org/#names-and-labels. I guess we'll need to add iconv sooner or later, but we should at least support all the aliases for the encodings we can do with node's internal encodings.

**Pull request closed without merging**